### ISSUE TITLE
Can't build project referencing 2.0.0-beta* on MacOS

### ISSUE THREAD

**Issue by georgiosd** [toddams/RazorLight#350]
On latest MacOS with dotnet core 3.1.5, a 3.1 project referencing `RazorLight` will fail with:

```
/Users/georgiosd/Documents/studoyo/api/xxxx.csproj : error NU1102: Unable to find package Microsoft.CodeAnalysis.CSharp with version (>= 3.3.0)
/Users/georgiosd/Documents/studoyo/api/xxxx.csproj : error NU1102:   - Found 62 version(s) in nuget.org [ Nearest version: 3.0.0 ]
/Users/georgiosd/Documents/studoyo/api/xxxx.csproj : error NU1102: Unable to find package Microsoft.CodeAnalysis.Common with version (>= 3.3.0)
/Users/georgiosd/Documents/studoyo/api/xxx.csproj : error NU1102:   - Found 62 version(s) in nuget.org [ Nearest version: 3.0.0 ]
    0 Warning(s)
    2 Error(s)
```

I tried downgrading all the way down to beta1 and it fails with a different error.

Any ideas?

**Comment by georgiosd** [comment]
Seems to be a problem with the package but if you know of any workarounds let me know!
https://github.com/dotnet/roslyn/issues/45516

**Comment by jzabroski** [comment]
This **may** be a duplicate of #347 .  In effect, we don't directly depend upon Microsoft.CodeAnalysis.CSharp or Microsoft.CodeAnalysis.Common.  Instead, we depend upon Microsoft.CodeAnalysis.Razor.  But, to be honest, debugging .NET link-load issues is complex because you have to travel all the way up the nuget stack to figure this stuff out.

![image](https://user-images.githubusercontent.com/447485/86015408-85b86d00-b9ef-11ea-8496-4c564e3c588e.png)

**Comment by georgiosd** [comment]
@jzabroski yes, I noticed all that. But then I removed RazorLight and added the package in question (which on nuget.org appear to have 3.6.0+ versions) and the dotnet core still couldn't find it so I think it's an issue with nuget or the package itself.

**Comment by jzabroski** [comment]
@georgiosd It looks like `nuget restore` supports a [Verbosity detailed](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-restore#options) flag.
Can you:
1. Output `nuget.exe` (no parameters) to dump the version info (I have 5.6.0.6591 on my work pc in Amazon Workspaces)
2. Output `nuget.exe restore -Verbosity detailed` (ignore for a second that you should restore via dotnet.exe)
3. Output `dotnet.exe restore --verbosity detailed`. It is possible `dotnet.exe restore --verbosity diagnostic` may provide even more info, though I don't know enough about Dotnet/Nuget internals to guess why dotnet.exe wouldn't just pass the same logging configuration through to nuget.exe underlying API.
4. Follow https://docs.microsoft.com/en-us/nuget/consume-packages/package-restore-troubleshooting - this step is pretty much a prerequisite to getting an issue escalated with Visual Studio / Nuget teams, as they literally say in this web page to follow these steps before submitting an issue.

**Comment by jzabroski** [comment]
@georgiosd I think this issue might only happen on Mac, which is why I can't reproduce it.  It may be a duplicate of #292 

I guess, my question is, how can I repro a Mac issue?  What's the best practice?  I don't own mac any more.  Mine fried two years ago and I went Lenovo/Windows 10 Pro due to WSL support letting me run Linux inside Windows.  One way would be to move to GitHub Actions for CI/CD and use a Mac host runner type.  https://devblogs.microsoft.com/dotnet/continuous-integration-workflow-template-for-net-core-desktop-apps-with-github-actions/  But I am curious as you also run your own software company, what your best practices are?

**Comment by jzabroski** [comment]
Googling for [NU1102 mac os x](https://www.google.com/search?q=NU1102+mac+os+x)

I found this: https://stackoverflow.com/questions/58074542/unable-to-find-package-microsoft-netcore-app-host-osx-x64-after-upgrading-to

**Comment by georgiosd** [comment]
Thanks for spending time on this @jzabroski !

We make mobile apps also, so while half our team owns a personal Mac, we found that it's often necessary to use an independent box and we have rented a dedicated mac mini for $99/mo. I think there was a cheaper option for $80/mo or so.

If it's something that can be solved with tests, you can definitely use a mac host - https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml

**Comment by georgiosd** [comment]
PS: I think it's unlikely to be the runtime because I only have 3.1.5 on my Mac right now, but I will check it

**Comment by jzabroski** [comment]
I think it's most likely the mac stuff wasn't packaged correctly by Microsoft (or the macos build stack is non-deterministic in some way).  I run another project, FluentMigrator, where I ran into huge issues where System.ComponentModel.DataAnnotations wasnt packaged by EF team in a way that would allow you to write netstandard2.0 assemblies against a 4.7.2 executable.  I notice that when packages are not maintained by the "in-box .NET" package set, they drift in packaging standards.  It's probably something similar.

Getting a mac mini or similar, not a bad idea.  I will check FB Marketplace for cheap refurb.  Is there a minimum OS i should run?

**Comment by georgiosd** [comment]
@jzabroski yeah if you can keep it at home it's obviously going to be cheaper! We have to have it accessible from anywhere so the online dedi box made sense. 

I dont know an awful lot about Mac OS versions but it seems that the earliest offered by MS is 10.14 - I wouldn't go any lower!

**Comment by jzabroski** [comment]
I should be able to get a Mac Mini on Sunday for $100 USD.  So, realistically, ETA to fix this is at least two weeks out.

**Comment by georgiosd** [comment]
Really? That cheap!? Wow. If you're just getting it for this fix, I can probably offer a temp account on our remote machine.

**Comment by jzabroski** [comment]
Yeah man.  FaceBook Marketplace.  I dont know if they have it in Greece but it's pretty sweet way to deal in secondary markets.  Sometimes you end up with crap but I'd say it's net saved me money.  (I once bought a Dr. Beats metronome on there that after I took it home, realized it didn't actually work correctly.)  This Mac Mini is 10.11 but I can use dosdude to upgrade to Mojave.

**Comment by georgiosd** [comment]
Very cool indeed! We certainly have it here but I don't use it very much because of quality concerns, but for that price, who can argue :)

Let me know what you find! I'm working from my laptop today so will have to uncomment RazorLight usages

**Comment by georgiosd** [comment]
So when adding the `Microsoft.CodeAnalysis.CSharp` package I noticed it then complains about `Microsoft.CodeAnalysis.Common` package. Then I tried installing that instead and I think the last line is telling:

```
error: Package 'Microsoft.CodeAnalysis.Common' is incompatible with 'all' frameworks in project
```

**Comment by georgiosd** [comment]
Another interesting fact. Adding RazorLight 1.x works fine. Adding `Microsoft.CodeAnalysis.Common` on a fresh project adds 3.0.0. Then adding RazorLight 2.0 fails because it wants > 3.3.0 which cannot be found :S

**Comment by jzabroski** [comment]
Until I have a Mac I don't think I can figure out what MSBuild/Nuget is doing to cause you so much headaches.

**Comment by georgiosd** [comment]
I know, just adding context :)

**Comment by jzabroski** [comment]
A dumb idea, but can you try `<RollForward>LatestFeature</RollForward>` and see if that changes the outcome?  It shouldn't, but hey...

**Comment by georgiosd** [comment]
Nope...

**Comment by jzabroski** [comment]
@georgiosd I think this is actually related to #360

**Comment by georgiosd** [comment]
@jzabroski how could that be true? My project is ASP.NET Core already :S

**Comment by georgiosd** [comment]
Hey @jzabroski - what's the status of this? I'm getting a similar error on Windows now with RC2!

```
Unhandled exception. System.AggregateException: Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: RazorLight.IEngineHandler Lifetime: Singleton ImplementationType: RazorLight.EngineHandler': No constructor for type 'RazorLight.EngineHandler' can be instantiated using services from the service container and default values.)
 ---> System.InvalidOperationException: Error while validating the service descriptor 'ServiceType: RazorLight.IEngineHandler Lifetime: Singleton ImplementationType: RazorLight.EngineHandler': No constructor for type 'RazorLight.EngineHandler' can be instantiated using services from the service 
container and default values.
 ---> System.InvalidOperationException: No constructor for type 'RazorLight.EngineHandler' can be instantiated using services from the service container and default values.
```

**Comment by jzabroski** [comment]
This appears to be a different error than the one you previously reported.

Can you please check how you are registering RazorLight with your container?

**Comment by georgiosd** [comment]
You're right, I mixed up the issues. I solved this one by doing `services.AddSingleton(Options.Create(engine.Options))`

**Comment by jzabroski** [comment]
@georgiosd As a check - is this original issue now also resolved on MacOS?  If yes, for what target frameworks did you confirm it now works.  Thank you.

**Comment by jzabroski** [comment]
@georgiosd Just wondering if your original issue is resolved?  It turns out github actions supports macos images for virtual environments!

https://github.com/actions/virtual-environments/tree/main/images/macos

**Comment by georgiosd** [comment]
Forgive me , I haven't checked yet, give me a few days! Happy new year!

**Comment by jzabroski** [comment]
@georgiosd I am going to close this now that I merged #407 which covers the build matrix for RazorLight #299

**Comment by georgiosd** [comment]
Well done! Happy new year.
